src/main/java/picard/illumina/IlluminaBasecallsToFastq.java

```diff
@@ -632,77 +508,138 @@ public void close() {
+                ReadType readType = ReadType.valueOf(readTypeString);
+
+                byte[] bases = new byte[numBases];
+                byte[] quals = new byte[numBases];
+
+                for (int j = 0; j < numBases; j++) {
+                    bases[j] = this.binaryCodec.readByte();
+                    quals[j] = this.binaryCodec.readByte();
+                }
+
+                ReadData read = new ReadData();
+                read.setReadType(readType);
+                read.setBases(bases);
+                read.setQualities(quals);
+
+                readData[i] = read;
+            }
+
+            boolean pf = this.binaryCodec.readBoolean();
+            String matchedBarcode = this.binaryCodec.readLengthAndString(false);
+
+            ClusterData clusterData = new ClusterData(readData);
+            clusterData.setLane(lane);
+            clusterData.setTile(tile);
+            clusterData.setX(x);
+            clusterData.setY(y);
+            clusterData.setPf(pf);
+            if (matchedBarcode.length() == 0) {
+                clusterData.setMatchedBarcode(null);
+            } else {
+                clusterData.setMatchedBarcode(matchedBarcode);
             }
+
+            return clusterData;
         }
 
         @Override
-        public FastqRecordsForCluster decode() {
-            if (!reader.hasNext()) return null;
-            final FastqRecordsForCluster ret = new FastqRecordsForCluster(numTemplates, numSampleBarcodes, numMolecularBarcodes);
-            decodeArray(ret.templateRecords);
-            decodeArray(ret.sampleBarcodeRecords);
-            decodeArray(ret.molecularBarcodeRecords);
-            return ret;
+        public SortingCollection.Codec<ClusterData> clone() {
+            return new ClusterDataCodec();
         }
+    }
 
-        private void decodeArray(final FastqRecord[] recs) {
-            for (int i = 0; i < recs.length; ++i) {
-                recs[i] = reader.next();
-            }
+    /**
+     * Comparator used for sorting `ClusterData` objects by read name.
+     */
+    private static class ClusterDataQueryNameComparator implements Comparator<ClusterData> {
+        private final ReadNameEncoder readNameEncoder;
+
+        /**
+         * Creates a `ClusterData` comparator used to sort by query name given an read name encoder.
+         *
+         * @param readNameEncoder The read name encoder used to generate the read names for the clusters.
+         */
+        public ClusterDataQueryNameComparator(ReadNameEncoder readNameEncoder) {
+            this.readNameEncoder = readNameEncoder;
         }
 
         @Override
-        public SortingCollection.Codec<FastqRecordsForCluster> clone() {
-            return new FastqRecordsForClusterCodec(numTemplates, numSampleBarcodes, numMolecularBarcodes);
+        public int compare(ClusterData o1, ClusterData o2) {
+            return readNameEncoder.generateReadName(o1, 1).compareTo(readNameEncoder.generateReadName(o2, 1));
```

Comment:
Ugh, kinda sucks to have to generate the read name each time, but I see why.  Maybe we should at least add a method to ReadNameEncoder to generate the name without the stuff after the space since that'd be a bit more efficient?

Reply:
The other option would be to have a variable in ClusterData to hold a read name, then set it and compare it?

Reply:
Since ClusterData holds more than one read we'd either have to generate them all or a prefix and then add to it based on the pairNumber, In this case we use the first pair for comparison only. So we could do that as well.

That would also mean more data written for the sorting collection when it spills to disk.